drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+#include "adc.h"
+#include <zephyr/dt-bindings/adc/b91-adc.h>
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_b91);
+
+#define DT_DRV_COMPAT telink_b91_adc
+
+/* ADC B91 defines */
+#define SIGN_BIT_POSITION          (13)
+#define AREG_ADC_DATA_STATUS       (0xf6)
+#define ADC_DATA_READY             BIT(0)
+
+#define CASE_DT_GPIO_TO_HAL_GPIO(hal_gpio) \
+	case DT_##hal_gpio:			   \
+		adc_pin = hal_gpio;		   \
+		break;
+
+#define CASE_ACK_TIME_TO_HAL_SAMPLE_CYC(n) \
+	case ADC_ACQ_TIME(ADC_ACQ_TIME_TICKS, n):  \
+		sample_cycl = ADC_SAMPLE_CYC_##n;  \
+		break;
+
+/* B91 ADC driver data */
+struct b91_adc_data {
+	struct adc_context ctx;
+	int16_t *buffer;
+	int16_t *repeat_buffer;
+	uint8_t differential;
+	uint8_t resolution_divider;
+	struct k_sem acq_sem;
+	struct k_thread thread;
+
+	K_THREAD_STACK_MEMBER(stack, CONFIG_ADC_B91_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+struct b91_adc_cfg {
+	adc_ref_vol_e vref;
+	adc_sample_freq_e sample_freq;
+	uint16_t ref_internal;
+};
+
+/* Validate ADC data buffer size */
+static int adc_b91_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t needed = sizeof(int16_t);
+
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+/* Validate ADC read API input parameters */
+static int adc_b91_validate_sequence(const struct adc_sequence *sequence)
+{
+	int status;
+
+	if (sequence->channels != BIT(0)) {
+		LOG_ERR("Only channel 0 is supported.");
+		return -ENOTSUP;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("Oversampling is not supported.");
+		return -ENOTSUP;
+	}
+
+	status = adc_b91_validate_buffer_size(sequence);
+	if (status) {
+		LOG_ERR("Buffer size too small.");
+		return status;
+	}
+
+	return 0;
+}
+
+/* Convert dts pin to B91 SDK pin */
+static adc_input_pin_def_e adc_b91_get_pin(uint8_t dt_pin)
+{
+	adc_input_pin_def_e adc_pin = NOINPUTN;
+
+	switch (dt_pin) {
+	FOR_EACH(CASE_DT_GPIO_TO_HAL_GPIO, (),
+		 ADC_GPIO_PB0, ADC_GPIO_PB1, ADC_GPIO_PB2, ADC_GPIO_PB3,
+		 ADC_GPIO_PB4, ADC_GPIO_PB5, ADC_GPIO_PB6, ADC_GPIO_PB7,
+		 ADC_GPIO_PD0, ADC_GPIO_PD1, ADC_VBAT)
```

Comment:
Originally it was with the regular case implementation and it was changed to FOR_EACH because of the review comment in this PR.
Fixed. Reverted back to the original implementation.